### ISSUE TITLE
Use ansistrano instead of capistrano

### ISSUE THREAD

**Issue by MaximeThoonsen** [theodo/theodo-stack-generator#165]
Hello,

On our project we have issues deploying the last version of api-platform with capistrano:

- Warming up the cache is not working (it seems to be because it's using the "--no-debug" flag)
- The env var is not correctly set up: it uses "dev" instead of prod and it's not a configuration problem, it's because the name of the var have changed
- Debug is difficult because we don't have the source
- we need to have the correct version of ruby installed which can be a struggle

In the other hand, I have used Ansistrano on my last 4 projects and I didn't have major issues + I can customize very precisely what is happening.

cc @benja-M-1

**Comment by benja-M-1** [comment]
> Warming up the cache is not working (it seems to be because it's using the "--no-debug" flag)
> The env var is not correctly set up: it uses "dev" instead of prod and it's not a configuration problem, it's because the name of the var have changed
Can you provide a screenshot of the result of the execution of the failing command we saw together?

> Debug is difficult because we don't have the source
This is actually wrong otherwise it wouldn't work :)
Sources are by default installed in the installed gem directory. To see where this directory is use ```gem environment```.
You can set it to wherever you want with bundler, for example: ```bundle config --local vendor/bundler```
This will install the ruby vendors in the vendor/bundler path in your project.

> we need to have the correct version of ruby installed which can be a struggle
Agreed

Can you provide a PR with what you came up with on your project?

**Comment by Remy-Luciani** [comment]
"Nothing is more dangerous than an idea, when it's the only one we have."

@MaximeThoonsen did you benchmark other solutions?

I used shipit to deploy React frontend apps and NodeJS backend app. It was OK, to maintain and use. But not declarative enough imo.

I don't code PHP anymore and never had the opportunity to test [Deployer](https://github.com/deployphp/deployer).

Other ideas: https://alternativeto.net/software/capistrano/ (Ansistrano is there).

We need some good criteria to choose the right tool. I suggest:

✓ fast to deploy the code generated the first time (< 2 mind)
✓ highly maintained repository
✓ awesome documentation
✓ easy to extend for custom tasks implementation
✓ no system dependecy overhead (new runtime, new package manager...)

What do you think?

**Comment by benja-M-1** [comment]
I love your criterias @Remy-Luciani  :) 

FYI here is an "official" list of tools from the Symfony documentation: https://symfony.com/doc/current/deployment.html#using-build-scripts-and-other-tools

**Comment by MaximeThoonsen** [comment]
@Remy-Luciani 
Oui j'ai essayé par le passé capistrano et shipit et j'avais regardé d'autres trucs mais qui n'étaient pas mieux que capistrano ou ansistrano.

✓ fast to deploy the code generated the first time (< 2 mind) => Si le générateur est à jour ce sera OK pour tous
✓ highly maintained repository => [last commit 6 days ago](https://github.com/ansistrano/deploy)
✓ easy to extend for custom tasks implementation => il n'y a que Ansistrano qui permette de modifier ce qu'il se passe à chaque étape. C'est vraiment son point fort
✓ no system dependecy overhead (new runtime, new package manager...) => besoin d'Ansible et de python qui sont déjà un requirement ATM VS ruby et son potentiel nightmare (j'ai déjà perdu quelques jours à installer/faire fonctionner capistrano)

On peut choisir un autre outils mais je me porte pas garant du fait que ça fonctionne directement :).

**Comment by MaximeThoonsen** [comment]
Deployer j'avais testé sur fundshop mais ça m'avait pas convaincu, je me rappelle plus pourquoi :/.

Après c'est bien de vouloir benchmarker la terre entière mais ça prend du temps....

**Comment by benja-M-1** [comment]
> We need some good criteria to choose the right tool. I suggest:
> 
> ✓ fast to deploy the code generated the first time (< 2 mind)
> ✓ highly maintained repository
> ✓ awesome documentation
> ✓ easy to extend for custom tasks implementation
> ✓ no system dependecy overhead (new runtime, new package manager...)
> 

@Remy-Luciani tu peux faire l'exercice pour shipit du coup ?

**Comment by Remy-Luciani** [comment]
@MaximeThoonsen 
> il n'y a que Ansistrano qui permette de modifier ce qu'il se passe à chaque étape. C'est vraiment son point fort

This is wrong. Capistrano has a bunch of events where you can add any task in a declarative way. Shipit is based on a task orchestrator with similar event workflow, a plugin system and its API allow you to write procedural task anyhow.

> Après c'est bien de vouloir benchmarker la terre entière mais ça prend du temps....

I agree. But imo, this generator is here to avoid all teams to waste time everyday, not once. I think we should not "Move fast, break things". We need more time and more people. I might have more time soon, count on me to help. ;)

__________

For ShipIt so:

✓ fast to deploy the code generated the first time (< 2 min)
Deployed Saham's MyAuto front + backend apps in less than 1'30" everyday on AWS Frankfurt datacenter from Casablanca
? highly maintained repository
Last commit the 14th of June, 2018. 4k+ stars. Les commit, releases and contributors than Ansistrano. I think the JS community does bring more stars. Last commir 3 months ago, is it "highly" maintained? Idk.
✓ awesome documentation
There is an API reference in the README and an example of a standard deploy
✓ easy to extend for custom tasks implementation
Plugin system, event based workflow, procedural task with bash command (erk) possible
✓ no system dependecy overhead (new runtime, new package manager...)
We use npm and nodejs to install this generator and React. No need to install NodeJS on the remote machine because everything is done through SSH.

**Comment by Remy-Luciani** [comment]
Seen with @MaximeThoonsen , Ansistrano seems to be the best solution so far:

- It is not a deployment automation _library_, but a bundle of pre-written Ansible playbook/roles used for automated deployment. Which means it can be harmlessly hacked for bespoken needs, which happens a lot on deployment matters.
- Debugging Ansible is easy (DEBUG mode, verbosity, because all things are a task, you can pin point where to look in a blink of an eye).
- I wanted to test Deployer. **But**, PHP does not have a virtualenv/nvm like tool and managing. Because PHP7 is not the standard runtime on all OS yet and we do have many legacy project using PHP 5.3+, the use of a PHP lib for frequent tasks is discouraged.

In other words: **Ansistrano Wins** 🏆

**Comment by JusteLeblanc** [comment]
@Remy-Luciani 👍 
@MaximeThoonsen would love coding on generator to add it with you as I will need to replace Shipit by Ansistrano on my future project.

It will be interesting to have a subgenerator to add Ansistrano on any existing project. @benja-M-1 agree with that ?

**Comment by JusteLeblanc** [comment]
@matthieuauger why did you close it as Cap is still what is used in Forge for deploying Sf apps ?
